[FLINK-1994] [ml] Add different gain calculation schemes to SGD

rawkintrevo opened the pull request:
Continuation of pull request 1384. 

Fixed long line issue, rebased on top of current master, and changed title per instructions. 

https://github.com/apache/flink/pull/1384

tillrohrmann commented:
There are still some scalastyle violations

```
error file=/home/travis/build/apache/flink/flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala message=File line length exceeds 100 characters line=56
error file=/home/travis/build/apache/flink/flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala message=File line length exceeds 100 characters line=221
error file=/home/travis/build/apache/flink/flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala message=File line length exceeds 100 characters line=224
```

You can check that locally by simply running `mvn clean verify`.

rawkintrevo commented:
That last fail looked like it happened in Kafka from something I never touched? E.g. random. Is there anyway to just retry the travis-ci build?

StephanEwen commented:
The Kafka tests are not written in a bullet proof way. It may happen that the Kafka test-cluster has a hickup, in which case the tests cannot succeed...

You can (force) push changes to your branch, that re-triggers the CI.

rawkintrevo commented:
So... looks like this passed. What do I do next?

chiwanpark commented on flink-staging/flink-ml/pom.xml:
```diff
@@ -80,7 +80,7 @@
 		<plugins>
 			<plugin>
-				<groupId>org.scala-tools</groupId>
+        			<groupId>org.scala-tools</groupId>
```
Please don't change indent.

rawkintrevo commented:
I think I really screwed this up... I was just trying to undo the changes to flink-staging/flink-ml/pom.xml

chiwanpark commented:
Hi @rawkintrevo, I think you should rebase your branch instead of merging master branch. Could you update your branch? Note that you should force push (`git push -f origin flink-1994`) after rebasing your branch.

rawkintrevo commented:
So do I need to do anything else for this to get merged?

chiwanpark commented:
Hi @rawkintrevo, I'm sorry about waiting you.

I have looked your pull request. Almost of changes are good but I have some few comments.

First, there are some meaningless changes such as indentation in `flink-ml/pom.xml` file and definition of `optimize` method of `GradientDescent` class. Please revert this changes.

Second, how about using `Enum` for the added parameter? Numbers cannot express meaning of optimization method.

Third, I think that we should apply this changes to documentation. Please add this content into the FlinkML documentation. :)

tillrohrmann commented on flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala:
```diff
@@ -54,14 +54,15 @@ abstract class GradientDescent extends IterativeSolver {
     data: DataSet[LabeledVector],
-    initialWeights: Option[DataSet[WeightVector]]): DataSet[WeightVector] = {
+    initialWeights: Option[DataSet[WeightVector]]
+                         ): DataSet[WeightVector] = {
```
Indentation:
```
override def optimize(
    data: ...
    initialWeights: ...)
  : DataSet[WeightVector] = {

}
```

rawkintrevo commented:
on your second point- I agree that Enum would be better for human readable variable setting, but if one was doing a grid parameter search it would be easier to search over a range of integers than having to establish an array of specifically named variables. Not significantly harder, but still slightly more tedious. 

Further, since I'm just learning scala- limited research/understanding leads me to believe 1) scala doesn't have an enum that works just like Java, and 2) changing the cases to more informative string names would achieve the desired goal? 

In the mean time working on documentation and reverting the indentation (which I thought I had already done), and thanks for bearing with me as I am still learning.

chiwanpark commented:
Hi @rawkintrevo, you can convert scala Enum to Int like following:

```scala
object Parameter extends Enumeration {
  type Parameter = Value
  val Param1, Param2, Param3, Param4 = Value
}

val convertedInt: Int = Parameter.Param1.id
```

Also, these values can be caught by pattern matching:

```scala
val param = getSomeEnumValues()

param match {
  case Parameter.Param1 => // blahblah
  case Parameter.Param2 => // blahblah
  case Parameter.Param3 => // blahblah
  case Parameter.Param4 => // blahblah
}
```

rawkintrevo commented:
still refactoring to enumeration, but wanted to toss this up in case anyone is watching- I added a decay parameter which significantly generalizes the Xu and Inverse Scaling Methods. Also I added some docs, but am open to suggestions, for example- once I figure out how to set the method with enumeration, I'll include a note?

chiwanpark commented:
I still would like to use enumeration because if we use string parameters, the user cannot check that the input parameter is valid. But if you have some problems to use enumeration, I'll modify your pull request before merging it.

Failing CI seems unrelated this PR. I'll check and test this.

rawkintrevo commented:
I can't get the enumeration thing to work. If you can modify it that would be awesome.  Thanks.

tillrohrmann commented on docs/libs/ml/optimization.md:
```diff
@@ -256,6 +271,79 @@ The full list of supported prediction functions can be found [here](#prediction-
+- $\eta_j$ is the step size on step $j$
+- $\eta_0$ is the initial step size
+- $\lambda$ is the regularization constant
+- $k$ is the decay constant
```
Where is decay explained?

tillrohrmann replied:
In which formula is `k` actually used?

tillrohrmann commented on docs/libs/ml/optimization.md:
```diff
@@ -256,6 +271,79 @@ The full list of supported prediction functions can be found [here](#prediction-
+            This is the `'optimal'` method of sklearn.  Chooses optimal initial $t_0 = \lambda \cdot eta_0$, based on Leon Bottou's [Learning with Large Data Sets ](http://leon.bottou.org/slides/largescale/lstut.pdf)
+          </p>
+        </td>
+        <td class="text-center">\eta_j = \frac{1}{\lambda \cdot (\frac{1}{\lambda \cdot eta_0 } +j -1) }</td>
```
math environment missing

tillrohrmann commented on docs/libs/ml/optimization.md:
```diff
@@ -256,6 +271,79 @@ The full list of supported prediction functions can be found [here](#prediction-
+Averaged Stochastic Gradient Descent](http://arxiv.org/pdf/1107.2490.pdf).
+          </p>
+        </td>
+        <td class="text-center">\eta_j = \lambda \cdot (1+ \lambda \cdot \eta_0 \cdot j^{\$tau} )</td>
```
math environment missing

tillrohrmann commented on docs/libs/ml/optimization.md:
```diff
@@ -256,6 +271,79 @@ The full list of supported prediction functions can be found [here](#prediction-
+            A very common method for determining the step size.
+          </p>
+        </td>
+        <td class="text-center">$\eta_j = \frac{\lambda}{j^{\tau}}$</td>
```
Maybe don't use `frac` here but instead `\lambda/ j^\tau`, because the exponent of `j` is rendered really small with `frac`.

tillrohrmann replied:
What is `tau` by the way?

tillrohrmann commented on flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/regression/MultipleLinearRegression.scala:
```diff
@@ -107,6 +107,11 @@ class MultipleLinearRegression extends Predictor[MultipleLinearRegression] {
     this
   }
 
+  def setOptimizationMethod(optimizationMethod: String): this.type = {
```
If you add options to an algorithm, then you have to adapt the code and online documentation.

tillrohrmann replied:
Furthermore, if you expose the optimization method, then you should also expose the decay parameter. But to be honest I'm not so sure whether this is the right way to do, because you would have to add to every algorithm every new parameter of the underlying solver. I guess it would be better to expose a parameter where you can set a non-default solver which allows you to set these things directly. What do you think?

chiwanpark replied:
@tillrohrmann How about creating a trait called `OptimizationMethod` which has a 'learningRate' method and using a derived class as a parameter?

tillrohrmann replied:
@chiwanpark, that could also be a solution. Then you could create an instance of a sub class with the respective parameters (e.g. the decay value). That makes also more sense, since not all optimization methods depend on the decay value. I've opened a PR where I did something similar for the abstraction of the actual calculation scheme (https://github.com/rawkintrevo/flink/pull/1). This should be easy to extend to include the decay value only for the `Xu` and `InvScaling` methods. Then we can remove the decay parameter from `IterativeSolver`.

tillrohrmann commented:
I had some minor comments. I also opened a PR against your branch @rawkintrevo which shows how one could solve the problem with the enumerations (https://github.com/rawkintrevo/flink/pull/1). Would be great if we can fix these last issues :-)

rawkintrevo commented:
For sure, definitely want to close the books on this one, and I appreciate all your help and patience as I have been learning. I will work on this tonight.

tillrohrmann commented:
:+1:

rawkintrevo commented:
Sorry for the multiple commits- It wasn't registering the changes here so I kept pushing... I can squash them if you want. 

I update the docs per your comments, added a brief description of the meaning of the decay rate and noticed a typo in Wu's method (forgot a negative sign in the exponent).

chiwanpark commented on docs/libs/ml/optimization.md:
```diff
@@ -276,6 +369,8 @@ val sgd = GradientDescentL1()
   .setRegularizationConstant(0.2)
   .setIterations(100)
   .setLearningRate(0.01)
+  .setOptimizationMethod("xu")
```
This line should be changed to `.setOptimizationMethod(LearningRateMethod.Xu)`.

tillrohrmann replied:
It should actually be `.setLearningRateMethod(LearningRateMethod.Xu)`.

chiwanpark commented:
@rawkintrevo Thanks for update! I still prefer using a learning method instance as parameter to remove method-specific parameter such as `decay`. But It is not mandatory.

Looks good to merge. :+1:

tillrohrmann commented:
Not squashing the commits here is the right way to go. We'll squash them once we merge it.

tillrohrmann commented on docs/libs/ml/optimization.md:
```diff
@@ -189,9 +187,26 @@ The following list contains a mapping between the implementing classes and the r
         </td>
       </tr>
+      <tr>
+        <td><strong>OptimizationMethod</strong></td>
```
This should be renamed to `LearningRateMethod`

tillrohrmann commented:
LGTM, will merge it then. I'll address @chiwanpark comment concerning the `decay` value while merging it. Thanks for your work @rawkintrevo :-)

chiwanpark commented:
@tillrohrmann Thanks! It is good news.
@rawkintrevo Thanks for contribution :+1:

rawkintrevo commented:
Thank you guys!
